include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,203 @@
+            uint32_t const workerIdx = threadIdx.x;
+
+            using complex_X = pmacc::math::Complex< float_X >;
+            using namespace pmacc::mappings::threads;
+            using namespace pmacc;
+            // Storage for positions in the beam coordinate system.
+            PMACC_SMEM(
+                acc,
+                positions,
+                memory::Array< float2_X, blockSize >
+            );
+            // Storage for (form factor density * beam intensity factor).
+            PMACC_SMEM(
+                acc,
+                densities,
+                memory::Array< float_X, blockSize >
+            );
+
+            uint32_t const linAccessBlockBegin =  blockIdx.x * blockSize;
+
+            ForEachIdx<
+                IdxConfig<
+                    blockSize,
+                    numWorkers
+                >
+            >{ workerIdx }(
+                [ & ](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    // Each thread reads one field value and saves it together
+                    // with its position, in UNIT_LENGTH, in the global domain.
+                    floatD_X fieldGlobalPosition;
+                    DataSpace< simDim > const cellPosition(
+                        DataSpaceOperations< simDim >::map(
+                            cellsGrid,
+                            linearIdx + linAccessBlockBegin
+                        )
+                    );
+
+                    DataSpace< simDim > const
+                        cellGlobalPosition( cellPosition + globalOffset );
+                    fieldGlobalPosition = precisionCast< float_X >
+                        ( cellGlobalPosition ) + fieldPos( )[ 0 ];
+                    fieldGlobalPosition *= cellSize.shrink< simDim >( );
+
+                    float_X density = densityBoxGPU( cellPosition )[ 0 ];
+                    // Save the cell position in the beam coordinate system.
+                    float3_X  position_b = probingBeam.coordinateTransform(
+                        currentStep,
+                        fieldGlobalPosition
+                    );
+                    float_X beamFactor = probingBeam( position_b );
+                    // Store the cell value of the function that has to be
+                    // Fourier transformed.
+                    densities[ linearIdx ] = density * beamFactor;
+                    // Store position in the beam comoving system
+                    // 3rd component is not needed anymore since q_z = 0.
+                    positions[linearIdx] = position_b.shrink< DIM2 >( );
+                    // Wait for all threads on the block to finish.
+                    __syncthreads( );
+
+                    // Calculate the density fourier transform:
+                    constexpr float_X meter2angstrom = 1e10;
+                    float_X dotqr;
+                    complex_X amplitude;
```

Comment:
(Not important) these two may be defined inside the following double-loop, since they are not required to hold a value between iterations.